src/tasks/dump.ts

```diff
@@ -0,0 +1,567 @@
+  // standard erc20 transfer. This is only set if there is a receiver and the
+  // specified toToken is also a token to be dumped.
+  transferToReceiver?: TransferToReceiver;
+}
+
+interface GetDumpInstructionInput {
+  dumpedTokens: string[];
+  toTokenAddress: string | undefined; // undefined defaults to native token (e.g., ETH)
+  user: string;
+  allowanceManager: string;
+  maxFeePercent: number;
+  hasCustomReceiver: boolean;
+  hre: HardhatRuntimeEnvironment;
+  network: SupportedNetwork;
+}
+async function getDumpInstructions({
+  dumpedTokens: inputDumpedTokens,
+  toTokenAddress,
+  user,
+  allowanceManager,
+  maxFeePercent,
+  hasCustomReceiver,
+  hre,
+  network,
+}: GetDumpInstructionInput): Promise<DumpInstructions> {
+  // todo: support dumping ETH by wrapping them
+  if (inputDumpedTokens.includes(BUY_ETH_ADDRESS)) {
+    throw new Error(
+      `Dumping the native token is not supported. Remove the ETH flag address ${BUY_ETH_ADDRESS} from the list of tokens to dump.`,
+    );
+  }
+
+  const dumpedTokens = Array.from(new Set(inputDumpedTokens));
+  const toToken =
+    toTokenAddress === undefined || toTokenAddress === BUY_ETH_ADDRESS
+      ? nativeToken(hre)
+      : await tokenDetails(toTokenAddress, hre);
+
+  let transferToReceiver: TransferToReceiver | undefined = undefined;
+  const sameTokenIndex = dumpedTokens.findIndex(
+    (token) => token === (toTokenAddress ?? BUY_ETH_ADDRESS),
+  );
+  if (sameTokenIndex !== -1) {
+    dumpedTokens.splice(sameTokenIndex, 1);
+    if (hasCustomReceiver) {
+      transferToReceiver = {
+        token: toToken,
+        amount: await balanceOf(toToken, user),
+      };
+    }
+  }
+
+  const computedInstructions: (DumpInstruction | null)[] = (
+    await promiseAllWithRateLimit(
+      dumpedTokens.map((tokenAddress) => async ({ consoleWarn }) => {
+        const token = await tokenDetails(tokenAddress, hre);
+        const [balance, approvedAmount]: [
+          BigNumber,
+          BigNumber,
+        ] = await Promise.all([
+          token.contract.balanceOf(user).then(BigNumber.from),
+          token.contract.allowance(user, allowanceManager).then(BigNumber.from),
+        ]);
+        if (balance.isZero()) {
+          consoleWarn(
+            `Dump request skipped for token ${displayName(
+              token,
+            )}. No balance for that token is available.`,
+          );
+          return null;
+        }
+        const sellToken = token.address;
+        const buyToken = isNativeToken(toToken)
+          ? WRAPPED_NATIVE_TOKEN_ADDRESS[network] // todo: replace WETH address with BUY_ETH_ADDRESS when services support ETH estimates
+          : toToken.address;
+        const kind = OrderKind.SELL;
+        const fee = await getFee({
+          sellToken,
+          buyToken,
+          kind,
+          amount: balance,
+          network,
+        });
+        const amount = balance.sub(fee);
+        if (amount.lte(constants.Zero)) {
```

Comment:
I know there are a lot of runtime parameters already but I wonder if it might be worth including a thread hold for this so that one can say whether they would rather not dump something if it's gonna cost half as much in fees. 


Edit: oh I see this come next and is the max fee percentage parameter.